src/BlockTypes/ProductCollection.php

```diff
@@ -22,4 +38,269 @@ class ProductCollection extends AbstractBlock {
+	 *
+	 * @return array
+	 */
+	private function get_custom_orderby_query( $orderby ) {
+		if ( ! in_array( $orderby, $this->custom_order_opts, true ) ) {
+			return array( 'orderby' => $orderby );
+		}
+
+		$meta_keys = array(
+			'popularity' => 'total_sales',
+			'rating'     => '_wc_average_rating',
+		);
+
+		return array(
+			'meta_key' => $meta_keys[ $orderby ],
+			'orderby'  => 'meta_value_num',
+		);
+	}
+
+	/**
+	 * Return or initialize $valid_query_vars.
+	 *
+	 * @return array
+	 */
+	private function get_valid_query_vars() {
+		if ( ! empty( $this->valid_query_vars ) ) {
+			return $this->valid_query_vars;
+		}
+
+		$valid_query_vars       = array_keys( ( new WP_Query() )->fill_query_vars( array() ) );
+		$this->valid_query_vars = array_merge(
+			$valid_query_vars,
+			// fill_query_vars doesn't include these vars so we need to add them manually.
+			array(
+				'date_query',
+				'exact',
+				'ignore_sticky_posts',
+				'lazy_load_term_meta',
+				'meta_compare_key',
+				'meta_compare',
+				'meta_query',
+				'meta_type_key',
+				'meta_type',
+				'nopaging',
+				'offset',
+				'order',
+				'orderby',
+				'page',
+				'post_type',
+				'posts_per_page',
+				'suppress_filters',
+				'tax_query',
+			)
+		);
+
+		return $this->valid_query_vars;
+	}
+
+	/**
+	 * Merge two array recursively but replace the non-array values instead of
+	 * merging them. The merging strategy:
+	 *
+	 * - If keys from merge array doesn't exist in the base array, create them.
+	 * - For array items with numeric keys, we merge them as normal.
+	 * - For array items with string keys:
+	 *
+	 *   - If the value isn't array, we'll use the value comming from the merge array.
+	 *     $base = ['orderby' => 'date']
+	 *     $new  = ['orderby' => 'meta_value_num']
+	 *     Result: ['orderby' => 'meta_value_num']
+	 *
+	 *   - If the value is array, we'll use recursion to merge each key.
+	 *     $base = ['meta_query' => [
+	 *       [
+	 *         'key'     => '_stock_status',
+	 *         'compare' => 'IN'
+	 *         'value'   =>  ['instock', 'onbackorder']
+	 *       ]
+	 *     ]]
+	 *     $new  = ['meta_query' => [
+	 *       [
+	 *         'relation' => 'AND',
+	 *         [...<max_price_query>],
+	 *         [...<min_price_query>],
+	 *       ]
+	 *     ]]
+	 *     Result: ['meta_query' => [
+	 *       [
+	 *         'key'     => '_stock_status',
+	 *         'compare' => 'IN'
+	 *         'value'   =>  ['instock', 'onbackorder']
+	 *       ],
+	 *       [
+	 *         'relation' => 'AND',
+	 *         [...<max_price_query>],
+	 *         [...<min_price_query>],
+	 *       ]
+	 *     ]]
+	 *
+	 *     $base = ['post__in' => [1, 2, 3, 4, 5]]
+	 *     $new  = ['post__in' => [3, 4, 5, 6, 7]]
+	 *     Result: ['post__in' => [1, 2, 3, 4, 5, 3, 4, 5, 6, 7]]
+	 *
+	 * @param array $base First array.
+	 * @param array $new  Second array.
+	 */
+	private function array_merge_recursive_replace_non_array_properties( $base, $new ) {
```

Comment:
While I understand your perspective on refactoring this method and potentially others, such as get_valid_query_vars, into a Utils class like Utils/BlocksWPQuery.php, I would suggest caution for now. 

Should the need for these methods arise in other block contexts in the future, it would certainly make sense to generalize them in a Utils class. However, until that requirement is more evident, keeping these functions within the ProductCollection block may contribute to code readability and maintainability by preserving context.

Let's continue to monitor our needs as the codebase evolves. Refactoring for wider reuse can be undertaken when the value and necessity of doing so become clearer.

What do you think?

Reply:
I won't push for that, it was just a suggestion and I'm ok with your reasoning! 👍